tests/validator/validator.cpp

```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessageValidator, validator->error(0)->description());
+}
+
+TEST(Validator, importInvalidUnitsNotDirectlyDeterminedFromImport)
+{
+    const std::string errorMessage =
+        "Imported units 'i_am_bad' is not valid because:\n"
+        "  -> Units 'ps2' imported as 'i_am_bad' from 'units_library.cellml' has an error:\n"
+        "  -> Prefix 'plinco' of a unit referencing 'second' in units 'ps2' is not a valid integer or an SI prefix.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/import_units_that_are_invalid.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto unitsLibrary = parser->parseModel(fileContents("importer/units_library.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    validator->validateModel(unitsLibrary);
+    EXPECT_EQ(size_t(2), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(errorMessage, validator->issue(0)->description());
+}
+
+TEST(Validator, importInvalidUnitsReference)
+{
+    const std::string errorMessage =
+        "Imported units 'i_am_bad' refers to units 'ps3' which does not appear in 'units_library'.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/import_units_that_are_not_there.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto unitsLibrary = parser->parseModel(fileContents("importer/units_library.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    validator->validateModel(unitsLibrary);
+    EXPECT_EQ(size_t(2), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(errorMessage, validator->issue(0)->description());
+}
+
+TEST(Validator, importInvalidComponentNotDirectlyDeterminedFromImport)
+{
+    const std::string errorMessage1 = "Component 'invalid_component' does not have a valid 'id' attribute, '8ioisfje+'.";
+    const std::string errorMessage2 =
+        "Imported component 'imported_component' is not valid because:\n"
+        "  -> Component 'invalid_component' imported as 'imported_component' from 'component_that_is_invalid.cellml' has an error:\n"
+        "  -> Component 'invalid_component' does not have a valid 'id' attribute, '8ioisfje+'.";
```

Comment:
I would rephrase to:
```
Component 'invalid_component' is imported from component 'imported_component' in 'component_that_is_invalid.cellml' which is invalid (it does not have a valid 'id' attribute ('8ioisfje+')).
```

(Double parentheses are not great, but I could personally live with them.)